platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/SyncTipsetEventCreationManager.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.event.tipset;
+
+import com.swirlds.platform.event.GossipEvent;
+import com.swirlds.platform.event.tipset.rules.TipsetEventCreationRule;
+import com.swirlds.platform.internal.EventImpl;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * Wraps a {@link TipsetEventCreator} and provides additional functionality. Will sometimes decide not to create new
+ * events based on external rules or based on paused status. Forwards created events to a consumer, and retries
+ * forwarding if the consumer is not immediately able to accept the event.
+ */
+public class SyncTipsetEventCreationManager {
```

Comment:
I split the business logic and threading apart to make the business logic simpler to unit test.